### ISSUE TITLE
Support direct iteration over heterogeneous tuples

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#14929]
Traditionally, Chapel has not supported direct iteration over heterogeneous tuples, like so:

```chapel
var tup = (1, 2.3, "hi");
for t in (1, 2.3, "hi") do
  writeln(tup);
```

I think one recent reason that we haven't considered supporting this is due to a sense that since `t` would have different types in different iterations, it simply wouldn't make sense (in fact, I argued this on issue #9607, which is strongly related to this one).  However, this argument doesn't really hold up in that it is at odds with how we typically describe the iteration variable in Chapel loops: namely that each iteration has its own personal copy of `t` (which is why things like forall loops and param loops work and make sense... they are not sharing a loop variable, but each iteration is logically getting its own).

As a result, users have commonly achieved the same result by writing the following pattern:

```chapel
var tup = (1, 2.3, "hi");
for param i in 1..tup.size do
  writeln(tup(i));
```

This works because the `param i` gives us the effect of fully unrolling the loop and getting a unique `i` value (and therefore unique `tup(i)` expression) per iteration.  However, as we try to rely less on features that expose 0- vs. 1-based indexing, such loops become less attractive (though even without such a change, the more direct form is also much more attractive).

For this reason, I'm proposing that Chapel should directly support iteration over heterogeneous tuples where the implementation will be understood to do the full unrolling of the loop to achieve it (essentially, to rewrite the first code snippet above into something like the following under the covers:

```chapel
{
  const ref i = tup(1);  // or tup(0) with 0-based tuples
  writeln(i);
}
{
  const ref i = tup(2);
  writeln(i);
}
{
  const ref i = tup(3);
  writeln(i);
}
```

).

[edit: As an extension to this, we could also support `[co]forall` loops over heterogeneous tuples that might transform the loop into a cobegin with a statement corresponding to each unrolled iteration, but this seems like it might have limited applicability to me, so for now I'm only focusing on the serial case].

**Comment by bradcray** [comment]
For a time, I'd been thinking this would be trivial to implement by simply writing the param-unrolled version of the loop into the tuple's iterator within ChapelTuple.chpl.  But I failed to anticipate that that doesn't work since iterators have to have a well-defined yield type.  So on further reflection, I think that this transformation will need to be done in the compiler.  If anyone familiar with handling of loops in the compiler (e.g., @vasslitvinov?) has suggestions about the best way to approach this, I'd be happy for suggestions.

**Comment by bradcray** [comment]
I'll note that I also thought that we should write such cases as:

```chapel
for param t in (1, 2.3, "hi") do
  ...
```

for a time, to leverage existing "param loop variables imply unrolling" logic and explanations.  But on reflection, this isn't appropriate / sufficient in all cases since the elements of the tuples may not be `param`.  For example, I think we can and should also support:

```chapel
config var i = 1,
                  r = 2.3,
                  s = "hi";
const tup = (i, r, s);
for t in tup do
  ...
```

but here `tup`'s elements aren't param values, so `for param t in tup do...` wouldn't make sense (`t` can't be a param since its values won't be known at compile-time).  Yet note that the rewrite in the OP above still works.  Thus, I think we just train people to understand that iteration over heterogeneous tuples results in unrolling.

**Comment by vasslitvinov** [comment]
Regarding implementation: the AST for a loop looks like this:
```
def _iterator
move _iterator, _getIterator(the Symbol for the tuple)
.....
ForLoop ( mIterator = SymExpr(_iterator), .... }
```

So I propose to modify `preFoldNamed()` in preFold.cpp to recognize a call to _getIterator on a heterogenous tuple and transform the ForLoop into whatever we want it to be, such as `for param i` .

**Comment by vasslitvinov** [comment]
I also propose to introduce for-expressions over tuples that produce tuples. Right now we use these workarounds for the lack of this feature:

* A recursive function that builds builds a bigger tuple from a smaller tuple. Besides bulky code, this pattern also repeats the core computation twice. Ex.:
```chpl
  inline proc _getIteratorZip(x: _tuple) {
    inline proc _getIteratorZipInternal(x: _tuple, param dim: int) {
      if dim == x.size then
        return (_getIterator(x(dim)),);
      else
        return (_getIterator(x(dim)), (..._getIteratorZipInternal(x, dim+1)));
    }
    if x.size == 1 then
      return _getIterator(x(1));
    else
      return _getIteratorZipInternal(x, 1);
  }
```

* Create a default-initialized tuple then fill it in a for-loop over the source tuple. This looks a bit prettier, although not ideal, yet relies on the ability to default-initialize. Ex.
```chpl
proc Block.targetLocsIdx(ind: rank*idxType) {
  var result: rank*int;
  for param i in 1..rank do
    result(i) = ......;
  return .....;
}
```

BTW here is an interesting extension of the previous pattern:
```chpl
iter Block.activeTargetLocales(const space : domain = boundingBox) {
  ...
  var dims : rank*range(low(1).type);
  for param i in 1..rank {
    dims(i) = low(i)..high(i);
  }
```

**Comment by bradcray** [comment]
That seems like it deserves its own issue rather than being glommed onto this one?

**Comment by vasslitvinov** [comment]
Good point. I created #14931.

**Comment by cassella** [comment]
What would the error message look like for invoking something like this on a heterogeneous tuple?

`iter foo(tup: _tuple) { for t in tup do yield t; }`

or

```chapel
proc foo(tup: _tuple, i: int) throws {
   for (t,c) in zip(tup,0..) { // Could you even do zippered iteration over it?
       if c == i then return t;
   }
   throw new Error(); 
}
```


I think it would be helpful to include in the error at least one pair of indices that result in differing return/yield types.

**Comment by bradcray** [comment]
I expect the error would be the same as any other iterator that tried to yield multiple distinct types, e.g.: https://tio.run/##S85ILEjN@f8/syS1SCEtP19DU6GaS0GhMjM1J0XB0BrONNIzRnCUMjKVrLlqubjS8osUMhUy86A6U/KBSsqLgGbl5Glkalr//w8A

I doubt we'll be able to support zippered iteration over heterogeneous tuples, at least in the initial implementation (zippering works by pushing things into a standard interface, and that standard interface would have similar challenges with yielding different types on different iterations; we could consider supporting zippering with other things that could be fully unrolled, but today unbounded ranges can't be, though with additional work they potentially could be as a follower.  I don't believe any form of fully unrolled loop supports zippering today (e.g., `for param i in 1..3`).

**Comment by bradcray** [comment]
[This](https://github.com/chapel-lang/chapel/issues/14929#issuecomment-588497981) was really helpful, thanks Vass.

**Comment by bradcray** [comment]
Have a 95% working solution on PR#15034.

**Comment by mppf** [comment]
Can we close this issue, then?

**Comment by bradcray** [comment]
Oops, yep, thanks!  Resolved by PR #15034.
